mbus-api/src/receive.rs

```diff
@@ -0,0 +1,158 @@
+use super::*;
+
+/// Type safe wrapper over a message bus message which decodes the raw
+/// message into the actual payload `S` and allows only for a response type `R`.
+///
+/// # Example:
+/// ```
+/// let raw_msg = &subscriber.next().await?;
+/// let msg: ReceivedMessage<RequestConfig, ReplyConfig> =
+///             raw_msg.try_into()?;
+///
+/// msg.respond(ReplyConfig {}).await.unwrap();
+/// // or we can also use the same fn to return an error
+/// msg.respond(Err(Error::Message("failure".into()))).await.unwrap();
+/// ```
+pub struct ReceivedMessage<'a, S, R> {
+    request: SendPayload<S>,
+    bus_message: &'a nats::asynk::Message,
+    reply_type: PhantomData<R>,
+}
+
+impl<'a, S, R> ReceivedMessage<'a, S, R>
+where
+    for<'de> S: Deserialize<'de> + 'a + Debug + Clone + MessageIdTrait,
+    R: Serialize,
+{
+    /// Get a copy of the actual payload data which was received.
+    pub fn payload(&self) -> S {
+        self.request.data.clone()
+    }
+    /// Get the sender identifier
+    pub fn sender(&self) -> SenderId {
+        self.request.sender.clone()
+    }
+
+    /// Respond back to the sender with the `reply` payload wrapped by
+    /// a Result-like type.
+    /// May fail if serialization of the reply fails or if the
+    /// message bus fails to respond.
+    /// Can receive either `R`, `Err()` or `ReplyPayload`.
+    pub async fn respond<T: Into<ReplyPayload<R>>>(
+        &self,
+        reply: T,
+    ) -> io::Result<()> {
+        let reply: ReplyPayload<R> = reply.into();
+        let payload = serde_json::to_vec(&reply)?;
```

Comment:
If the ReplyPayload would have the guard Serialize -- the conversion can be done in one blow?

Reply:
hmm but then it depends on what you pass on to the fn. Eg, If you call the fn with R rather than Ok(R) then we'd serialize R because we didn't do the .into to the ReplyPayload<R>?